Benchmarks/Benchmarks/Predicates/Predicates.swift

```diff
@@ -0,0 +1,234 @@
+                (monster.level == 80)
+            }
+
+            benchmark.startMeasurement()
+            var matched = 0
+            for _ in benchmark.scaledIterations {
+                if try predicate.evaluate(monster) {
+                    matched += 1
+                }
+            }
+            benchmark.stopMeasurement()
+
+            guard matched == benchmark.scaledIterations.count else {
+                fatalError("Internal error: wrong number of matched monsters")
+            }
+        }
+    }
+
+    if #available(macOS 14, *) {
+        Benchmark("Predicate #3 - 1 KeyPath computed property condition") { benchmark in
+            let monster = Monster(name: "Orc", level: 80, hp: 100, mana: 0, weapon: .sword(Sword(p1: 1, p2: 2, p3: 3, p4: 4, p5: 5)))
+            let predicate = #Predicate<Monster> { monster in
+                (monster.levelComputed == 80)
+            }
+
+            benchmark.startMeasurement()
+            var matched = 0
+            for _ in benchmark.scaledIterations {
+                if try predicate.evaluate(monster) {
+                    matched += 1
+                }
+            }
+            benchmark.stopMeasurement()
+
+            guard matched == benchmark.scaledIterations.count else {
+                fatalError("Internal error: wrong number of matched monsters")
+            }
+        }
+    }
+
+    if #available(macOS 14, *) {
+        Benchmark("Predicate #4 - 1 KeyPath nested computed property condition") { benchmark in
+            let monster = Monster(name: "Orc", level: 80, hp: 100, mana: 0, weapon: .sword(Sword(p1: 1, p2: 2, p3: 3, p4: 4, p5: 5)))
+            let predicate = #Predicate<Monster> { monster in
+                (monster.weaponP1 == 1)
+            }
+
+            benchmark.startMeasurement()
+            var matched = 0
+            for _ in benchmark.scaledIterations {
+                if try predicate.evaluate(monster) {
+                    matched += 1
+                }
+            }
+            benchmark.stopMeasurement()
+
+            guard matched == benchmark.scaledIterations.count else {
+                fatalError("Internal error: wrong number of matched monsters")
+            }
+        }
+    }
+
+    if #available(macOS 14, *) {
+        Benchmark("Predicate #5 - 3 KeyPath nested computed property conditions") { benchmark in
+            let monster = Monster(name: "Orc", level: 80, hp: 100, mana: 0, weapon: .sword(Sword(p1: 1, p2: 2, p3: 3, p4: 4, p5: 5)))
+            let predicate = #Predicate<Monster> { monster in
+                ((monster.weaponP1 == 1) &&
+                 //(monster.weaponP2 == 2) &&
+                 //(monster.weaponP3 == 3) &&
+                 (monster.weaponP4 == 4) &&
+                 (monster.weaponP5 == 5))
+            }
+
+            benchmark.startMeasurement()
+            var matched = 0
+            for _ in benchmark.scaledIterations {
+                if try predicate.evaluate(monster) {
```

Comment:
This logic is repeated a bit throughout this file, to make it easier to write predicate benchmarks in the future, maybe we could extract this out to some

```
func measureEvaluation<T>(_ predicate: Predicate<T>, input: T, benchmark: Benchmark)
```

Or similar so that each of these benchmarks could be as simple as creating the predicate and then just calling `measureEvaluation` which handles starting/stopping the measurement and calling evaluate the right number of times. It'd be nice to use that to create a lot more benchmarks quickly in the future. Thoughts?

Reply:
I think we can define a static array of predicates that should be tested and just have a single body, even without resorting to generics - I'll have a look next week when back in office.